integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaScalingTests.java

```diff
@@ -0,0 +1,142 @@
+package com.sequenceiq.it.cloudbreak.testcase.e2e.freeipa;
+
+import static com.sequenceiq.it.cloudbreak.context.RunningParameter.key;
+
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testng.annotations.Test;
+
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.FormFactor;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.Status;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.instance.InstanceGroupResponse;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.instance.InstanceMetaDataResponse;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.instance.InstanceMetadataType;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.describe.DescribeFreeIpaResponse;
+import com.sequenceiq.it.cloudbreak.FreeIpaClient;
+import com.sequenceiq.it.cloudbreak.client.FreeIpaTestClient;
+import com.sequenceiq.it.cloudbreak.context.Description;
+import com.sequenceiq.it.cloudbreak.context.TestContext;
+import com.sequenceiq.it.cloudbreak.dto.freeipa.FreeIpaDownscaleTestDto;
+import com.sequenceiq.it.cloudbreak.dto.freeipa.FreeIpaTestDto;
+import com.sequenceiq.it.cloudbreak.dto.freeipa.FreeIpaUpscaleTestDto;
+import com.sequenceiq.it.cloudbreak.dto.telemetry.TelemetryTestDto;
+import com.sequenceiq.it.cloudbreak.exception.TestFailException;
+import com.sequenceiq.it.cloudbreak.testcase.e2e.AbstractE2ETest;
+
+public class FreeIpaScalingTests extends AbstractE2ETest {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FreeIpaScalingTests.class);
+
+    private static final Status FREEIPA_AVAILABLE = Status.AVAILABLE;
+
+    @Inject
+    private FreeIpaTestClient freeIpaTestClient;
+
+    @Test(dataProvider = TEST_CONTEXT)
+    @Description(
+            given = "there is a running cloudbreak",
+            when = "a valid stack create request is sent with 1 FreeIPA instance " +
+                    "AND the stack is scaled up to HA " +
+                    "AND the stack is scaled down to TWO_NODE_BASED " +
+                    "AND the stack is scaled up to HA ",
+            then = "the stack should be available AND deletable and have 3 nodes AND the primary gateway should not change")
+    public void testFreeIpaUpAndDownscale(TestContext testContext) {
+        String freeIpa = resourcePropertyProvider().getName();
+        Set<String> primaryGatewayInstanceId = new HashSet<>();
```

Comment:
minor: why is this a `Set` when we _always_ expect to have only 1 pgw to exist?

Reply:
That is just because of the lambda quasi const requirement: I needed a container class, that is mutable. If there is a class that can hold at most 1 instance and is modifiable, then I would be happy to change Set to that.